### PR TITLE
[Impeller] add basic culling checks during text frame dispatcher.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/155133

Dl dispatching still relies on cull rects computed during that dispatch process. Make sure that the text frame dispatcher doesn't populate text frames that are way offscreen.

This culling is more conservative than the rendering dispatcher. We'd need to do some refactoring so the logic isn't repeated multiple times.

### CODE REVIEW

File: impeller/display_list/dl_dispatcher.h [hunk 1]
```diff
@@ -383,9 +385,12 @@ class TextFrameDispatcher : public flutter::IgnoreAttributeDispatchHelper,
   void setStrokeJoin(flutter::DlStrokeJoin join) override;
 
  private:
+  const Rect GetCurrentLocalCullingBounds() const;
+
   const ContentContext& renderer_;
   Matrix matrix_;
   std::vector<Matrix> stack_;
+  std::vector<Rect> cull_rect_state_;
```
Comment (flar) [inline impeller/display_list/dl_dispatcher.h:393]: I don't think you need a stack. At least not yet. SaveLayer bounds culling already happened so using a stack so that you can incorporate that into the culling of DL dispatching isn't likely to benefit much.

DL already culled - by failing to record - anything that was distinctly outside of a clip set in the DL stream, including the bounds of SaveLayers. So trying to intersect the cull_rect as you read through the current one isn't likely to improve the culling much compared to just culling against the (inverse transformed) frame bounds. If you remove the code in SaveLayer does it change the results much for your test case? If not then you don't really need a stack here until we do a full clip-based culling and even that would be icing on the cake rather than a critical step in the process.

File: shell/common/snapshot_controller_impeller.cc [hunk 2]
```diff
@@ -73,8 +73,11 @@ sk_sp<DlImage> DoMakeRasterSnapshot(
     );
   }
 
-  impeller::TextFrameDispatcher collector(context->GetContentContext(),
-                                          impeller::Matrix());
+  impeller::TextFrameDispatcher collector(
+      context->GetContentContext(),                 //
+      impeller::Matrix(),                           //
+      impeller::Rect::MakeSize(render_target_size)  //
+  );
```
Comment (flar) [inline shell/common/snapshot_controller_impeller.cc:80]: This pattern of "first dispatch to a TFD and then to a DlDispatcher" happens in a number of places. It feels like a convenience method might be nice.

**Review by flar — COMMENTED** [review]
Just a quick look while I'm OOO, but I had a couple of thoughts to consider...

File: impeller/display_list/dl_dispatcher.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/display_list/dl_dispatcher.h:393]: We do still need to account for image filters, which aren't captured in the transform state. The locally transformed device bounds may be too small in some cases.

Comment (flar) [reply impeller/display_list/dl_dispatcher.h:393]: Are you sure? I don't think regular display list culling takes those into account. It just forwards the inverse-mapped clip bounds to the new Dispatch call.

Keep in mind that the bounds recorded in the RTree include any expansion due the filters applied to them, even including every filter on every saveLayer that they are inside of. That kind of inverts the responsibility there, but I'd have to look at it more to be sure this is right. If it is wrong then maybe we're doing the rendering side of culling wrong as well...

Comment (jonahwilliams) [reply impeller/display_list/dl_dispatcher.h:393]: I'm sure the culling takes that into account, but this is more of a problem with the cull rect computation right? (assuming no other canvas ops) Suppose I have a NxM device rect and a saveLayer with a 2x scaling image filter on it. If I need to dispatch to a sub display list inside of that savelayer, It wouldn't be correct to use the NxM cull rect, I'd need a 2Nx2M cull rect.

For other transforms I can invert the transform matrix but we don't track the image filter transforms easily yet.

File: shell/common/snapshot_controller_impeller.cc (hunk 2 above)
Comment (jonahwilliams) [reply shell/common/snapshot_controller_impeller.cc:80]: In https://github.com/flutter/engine/pull/55019 I refactored this dispatching into a reusable static method.

File: impeller/display_list/dl_dispatcher.h (hunk 1 above)
Comment (flar) [reply impeller/display_list/dl_dispatcher.h:393]: Within a display list that has already been taken care of by the way that the RTree rects are accumulated, but I think you are correct that it is as if there is an additional "transform" that goes on when you hit a saveLayer that has a filter. So, just as you have to take the original cull rect and un-transform it by the CTM, you also have to un-transform it by the actions of the SaveLayers for similar reasons.

So, yes, it is still an issue that this code punts on the filter un-transform of the savelayers as you mentioned in the code.

File: impeller/display_list/dl_dispatcher.cc [hunk 3]
```diff
@@ -1368,22 +1368,40 @@ Canvas& ExperimentalDlDispatcher::GetCanvas() {
...
+  cull_rect_state_.push_back(cull_rect_state_.back());
 }
 
 void TextFrameDispatcher::saveLayer(const DlRect& bounds,
                                     const flutter::SaveLayerOptions options,
                                     const flutter::DlImageFilter* backdrop) {
   save();
+
+  // This dispatcher does not track enough state to accurately compute
+  // cull rects with image filters.
```
Comment (flar) [inline impeller/display_list/dl_dispatcher.cc:1388]: For this, and other reasons, I think saveLayer should be a NOP.

- First, DL has already done culling against things that appeared within the DL stream when it did the recording. So, the bounds you see here don't necessarily add any new information. The DL dispatch method is already culling things against the cull bounds that were provided in the initial dispatch call. DlBuilder already culled things against the bounds of save layers and clips when it did the recording. So, how is this particular enhancement really adding any value - other than some rare interactions where something might intersect the bounds here and also intersect the cull rect, but somehow not intersect the intersection of the two? If we were testing full geometries then strange cases like that might fall through the holes and we might be able to catch them with some more diligence here, but we are just testing bounds of the geometries, making it less likely to catch a corner case.
- You could more easily do this same kind of calculation in the clip*() methods which don't have the fuzzy issues of saveLayer. Why choose saveLayer to enforce its possibly-fuzzily-potentially clipping when we aren't instrumenting the actual clip calls?
- Really the important thing to track is how the filter manipulates the transform, but we aren't doing that. If anything, we should be setting the cull rect state to MaxRect here (like we have done in the non-experimental AIKS canvas) if there is a filter present, but that is not what the code here is doing...

I'd rather see this method just do:

```
  if (filter) {
    cull_rect_state.back() = Maximum;
  }
```

File: impeller/display_list/dl_dispatcher.cc [hunk 4]
```diff
@@ -1368,22 +1368,40 @@ Canvas& ExperimentalDlDispatcher::GetCanvas() {
...
 void TextFrameDispatcher::save() {
   stack_.emplace_back(matrix_);
+  cull_rect_state_.push_back(cull_rect_state_.back());
 }
 
 void TextFrameDispatcher::saveLayer(const DlRect& bounds,
                                     const flutter::SaveLayerOptions options,
                                     const flutter::DlImageFilter* backdrop) {
   save();
+
```
[the remark this answers is not in the document]
Comment (flar) [reply impeller/display_list/dl_dispatcher.cc:1386]: Also, your cull rect state doesn't look balanced. You always pop it in restore, but you push_back here. If save() doesn't push_back then restore will unbalance it. If save() does do a stack dup, then this code will over balance it...

**Review by flar — CHANGES_REQUESTED** [review]

File: impeller/display_list/dl_dispatcher.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/dl_dispatcher.cc:1386]: My understanding was that the cull rect always need to be in local (to the display list) coordinate space. So I need to transform the cull rect into local space and then intersect with the bounds.

File: impeller/display_list/dl_dispatcher.cc [hunk 5]
```diff
@@ -1368,22 +1368,40 @@ Canvas& ExperimentalDlDispatcher::GetCanvas() {
...
+  // This dispatcher does not track enough state to accurately compute
+  // cull rects with image filters.
+  if (paint_.image_filter) {
+    cull_rect_state_.push_back(bounds);
+  } else {
+    auto new_cull_rect = GetCurrentLocalCullingBounds().Intersection(bounds);
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/dl_dispatcher.cc:1402]: but `bounds` is in the local coordinate space, so I need the device cull rect transformed into local coordinates?

File: impeller/display_list/dl_dispatcher.cc (hunk 3 above)
Comment (jonahwilliams) [reply impeller/display_list/dl_dispatcher.cc:1398]: Changed this to set to max cull rect if an image filter is encountered.

File: impeller/display_list/dl_dispatcher.cc [hunk 6]
```diff
@@ -1368,22 +1368,52 @@ Canvas& ExperimentalDlDispatcher::GetCanvas() {
...
 
-void TextFrameDispatcher::save() {
+TextFrameDispatcher::~TextFrameDispatcher() {
+  FML_DCHECK(cull_rect_state_.size() == 1);
+}
+
+void TextFrameDispatcher::Save(bool push_cull_rect) {
   stack_.emplace_back(matrix_);
+  if (push_cull_rect) {
+    cull_rect_state_.push_back(cull_rect_state_.back());
```
Comment (flar) [inline impeller/display_list/dl_dispatcher.cc:1384]: An easier way to do this would be to have Save() always push it, and then SL replaces the value on ToS (`.back() = foo`) when it updates it?

File: impeller/display_list/dl_dispatcher.cc [hunk 7]
```diff
@@ -1459,16 +1489,40 @@ void TextFrameDispatcher::drawTextFrame(
...
-  save();
+  Save(/*push_cull_rect=*/true);
   Paint old_paint = paint_;
   paint_ = Paint{};
-  display_list->Dispatch(*this);
+  bool old_has_image_filter = has_image_filter_;
+  has_image_filter_ = false;
+
+  auto local_cull_bounds = GetCurrentLocalCullingBounds();
+  if (!local_cull_bounds.IsMaximum()) {
```
Comment (flar) [inline impeller/display_list/dl_dispatcher.cc:1512]: if IsEmpty then skip it...?

**Review by flar — APPROVED** [review]
LGTM - suggestions...

File: impeller/display_list/dl_dispatcher.cc (hunk 6 above)
Comment (jonahwilliams) [reply impeller/display_list/dl_dispatcher.cc]: Done. Testing now

File: impeller/display_list/dl_dispatcher.cc (hunk 7 above)
Comment (jonahwilliams) [reply impeller/display_list/dl_dispatcher.cc]: Done
